lib/kb_SetUtilities/kb_SetUtilitiesImpl.py

```diff
@@ -3070,6 +3735,762 @@ def KButil_Batch_Create_GenomeSet(self, ctx, params):
+                        'MT_Bioelement:Cfix': 'BioE:C_fixation',
+                        'MT_Bioelement:O': 'BioE:Oxygen',
+                        'MT_Bioelement:H': 'Bi0E:Hydrogen',
+                        'MT_Bioelement:NH3': 'BioE:Ammonia',
+                        'MT_Bioelement:CH4': 'BioE:CH4',
+                        'MT_Bioelement:S': 'BioE:Sulfur',
+                        'MT_Bioelement:Nit': 'BioE:Nit_Ox_Red'
+        }
+        fields = []
+        if int(params.get('show_sci_name',1)) == 1:
+            fields.append('sci_name')
+        fields.append('taxonomy')
+        if int(params.get('add_qc', 0)) == 1:
+            fields.extend(['qc_complete', 'qc_contam'])
+        fields.extend(['contig_count',
+                       'genome_length',
+                       'N50',
+                       'GC',
+                       'CDS_count',
+                       'tRNA_count',
+                       '5S_rRNA_count',
+                       '16S_rRNA_count',
+                       '23S_rRNA_count',
+                       'CRISPR_array_count'])
+        TSV_fields = []
+        TSV_fields.extend(fields)
+        if int(params.get('add_bioelement',0)) == 1:
+            TSV_fields.extend(['MT_Bioelement:Nfix',
+                               'MT_Bioelement:Cfix',
+                               'MT_Bioelement:O',
+                               'MT_Bioelement:H',
+                               'MT_Bioelement:NH3',
+                               'MT_Bioelement:CH4',
+                               'MT_Bioelement:S',
+                               'MT_Bioelement:Nit'])
+
+        # header
+        TSV_row = ['Genome']
+        for field in TSV_fields:
+            TSV_row.append(field_titles[field])
+        TSV_table_buf.append("\t".join(TSV_row))
+
+        # data
+        for genome_newVer_ref in genome_newVer_ref_order:
+            genome_obj_name = genome_newVer_ref_to_obj_name[genome_newVer_ref]
+            TSV_row = [genome_obj_name]
+            for field in TSV_fields:
+                TSV_row.append(table[genome_newVer_ref][field])
+            TSV_table_buf.append("\t".join(TSV_row))
+
+        # report in log
+        for row in TSV_table_buf:
+            self.log(console, row)
+
+        # write TSV to file and upload
+        TSV_file = 'GenomeSet_summary.tsv'
+        TSV_path = os.path.join(outdir, TSV_file)
+        TSV_str = "\n".join(TSV_table_buf)
+        with open(TSV_path, 'w') as TSV_handle:
+            TSV_handle.write(TSV_str)
+        try:
+            TSV_file_save_info = self.dfuClient.file_to_shock({'file_path': TSV_path, 'make_handle': 0})
+        except:
+            raise ValueError ("error saving TSV_file")
+        file_links = [{'shock_id': TSV_file_save_info['shock_id'],
+                       'name': TSV_file,
+                       'label': 'GenomeSet summary TSV'}]
+        
+        
+        #### build HTML table
+        ##
+        html_file_path = None
```

Comment:
I feel like there is a better way to do a lot of this using some templates and maybe CSS.  It must have been pretty tedious to iterate on this.

Reply:
yeah.  I should have used a template.